common/client/poller.go

```diff
@@ -15,83 +14,80 @@ import (
 // and delivers the result to a channel. It is used by multinode to poll
 // for new heads and implements the Subscription interface.
 type Poller[T any] struct {
-	services.StateMachine
+	services.Service
+	eng *services.Engine
+
 	pollingInterval time.Duration
 	pollingFunc     func(ctx context.Context) (T, error)
 	pollingTimeout  time.Duration
-	logger          logger.Logger
 	channel         chan<- T
 	errCh           chan error
-
-	stopCh services.StopChan
-	wg     sync.WaitGroup
 }
 
 // NewPoller creates a new Poller instance and returns a channel to receive the polled data
 func NewPoller[
 	T any,
-](pollingInterval time.Duration, pollingFunc func(ctx context.Context) (T, error), pollingTimeout time.Duration, logger logger.Logger) (Poller[T], <-chan T) {
+](pollingInterval time.Duration, pollingFunc func(ctx context.Context) (T, error), pollingTimeout time.Duration, lggr logger.Logger) (Poller[T], <-chan T) {
 	channel := make(chan T)
-	return Poller[T]{
+	p := Poller[T]{
 		pollingInterval: pollingInterval,
 		pollingFunc:     pollingFunc,
 		pollingTimeout:  pollingTimeout,
 		channel:         channel,
-		logger:          logger,
 		errCh:           make(chan error),
-		stopCh:          make(chan struct{}),
-	}, channel
+	}
+	p.Service, p.eng = services.Config{
+		Name:  "Poller",
+		Start: p.start,
+		Close: p.close,
+	}.NewServiceEngine(lggr)
+	return p, channel
 }
 
 var _ types.Subscription = &Poller[any]{}
 
-func (p *Poller[T]) Start() error {
-	return p.StartOnce("Poller", func() error {
-		p.wg.Add(1)
-		go p.pollingLoop()
-		return nil
-	})
+func (p *Poller[T]) start(ctx context.Context) error {
+	p.eng.Go(p.pollingLoop)
+	return nil
 }
 
 // Unsubscribe cancels the sending of events to the data channel
 func (p *Poller[T]) Unsubscribe() {
-	_ = p.StopOnce("Poller", func() error {
-		close(p.stopCh)
-		p.wg.Wait()
-		close(p.errCh)
-		close(p.channel)
-		return nil
-	})
+	_ = p.Close()
+}
+
+func (p *Poller[T]) close() error {
+	close(p.errCh)
+	close(p.channel)
+	return nil
```

Comment:
Seems like `Unsubscribe()` was added  instead of `Close()`because `services.Service` wasn't implemented?

Reply:
Does it make sense to replace it?

Reply:
We need a `func() error` to use the helper, but I guess it could be organized a different way :shrug: